From patchwork Mon Oct  8 20:10:17 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Kirti Wankhede <kwankhede@nvidia.com>
X-Patchwork-Id: 10631385
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id AB87C15E2
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Oct 2018 20:11:58 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 9FBD529C9B
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Oct 2018 20:11:58 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 9D35129D06; Mon,  8 Oct 2018 20:11:58 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham
	version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 1F5EC29CEE
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  8 Oct 2018 20:11:57 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726668AbeJIDZY (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 8 Oct 2018 23:25:24 -0400
Received: from hqemgate14.nvidia.com ([216.228.121.143]:5090 "EHLO
        hqemgate14.nvidia.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726393AbeJIDZY (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 8 Oct 2018 23:25:24 -0400
Received: from hqpgpgate101.nvidia.com (Not Verified[216.228.121.13]) by
 hqemgate14.nvidia.com (using TLS: TLSv1.2, DES-CBC3-SHA)
        id <B5bbbba070000>; Mon, 08 Oct 2018 13:11:51 -0700
Received: from HQMAIL101.nvidia.com ([172.20.161.6])
  by hqpgpgate101.nvidia.com (PGP Universal service);
  Mon, 08 Oct 2018 13:11:55 -0700
X-PGP-Universal: processed;
        by hqpgpgate101.nvidia.com on Mon, 08 Oct 2018 13:11:55 -0700
Received: from HQMAIL102.nvidia.com (172.18.146.10) by HQMAIL101.nvidia.com
 (172.20.187.10) with Microsoft SMTP Server (TLS) id 15.0.1395.4; Mon, 8 Oct
 2018 20:11:55 +0000
Received: from HQMAIL107.nvidia.com (172.20.187.13) by HQMAIL102.nvidia.com
 (172.18.146.10) with Microsoft SMTP Server (TLS) id 15.0.1395.4; Mon, 8 Oct
 2018 20:11:54 +0000
Received: from kwankhede-dev.nvidia.com (172.20.13.39) by HQMAIL107.nvidia.com
 (172.20.187.13) with Microsoft SMTP Server (TLS) id 15.0.1395.4 via Frontend
 Transport; Mon, 8 Oct 2018 20:11:52 +0000
From: Kirti Wankhede <kwankhede@nvidia.com>
To: <alex.williamson@redhat.com>, <eskultet@redhat.com>,
        <kvm@vger.kernel.org>, <libvir-list@redhat.com>
CC: <cjia@nvidia.com>, Kirti Wankhede <kwankhede@nvidia.com>
Subject: [PATCH v2 1/1] Add attribute single_usage_restriction for mdev
 type-id
Date: Tue, 9 Oct 2018 01:40:17 +0530
Message-ID: <1539029417-19371-1-git-send-email-kwankhede@nvidia.com>
X-Mailer: git-send-email 2.7.0
X-NVConfidentiality: public
MIME-Version: 1.0
Content-Type: text/plain
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=nvidia.com; s=n1;
        t=1539029511; bh=QVAitbw8hqlPAVPvmZdDxB3UD56hHqnShe669ciedRM=;
        h=X-PGP-Universal:From:To:CC:Subject:Date:Message-ID:X-Mailer:
         X-NVConfidentiality:MIME-Version:Content-Type;
        b=A5xwZnUHJ8ByxdFVw+QJE4xkkY6o5m5sMDA/hHh8OwuqcEHaz9r3EqPUeUUpVi84d
         q7v+TBhPoZMhDsHo5hTqurx7N+VYcjwRNYZcStxo+Hmw8ObTiu1oMbb/7jomQ5r0iE
         vmEUjI7FZ30ygIbH6CLFlLJ7u4WkAfLGI67i0R92wE5Xdfd5qka98AK+Rn7VLHBVgA
         IHnOk2Sgs65MX1O5AjQoT9uZImoGo0MSKp5dMIJenOhQ6mGI3DP0iof/F1BG144pYX
         5fhgnbHRGz6PeSbeCbNhvcWBXGz3uMMKkkqLPEXHMbNaZKGJjOPL2Ha6Te/ZXMhO0I
         BodjI+3oTgdbw==
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Generally a single instance of mdev device, a share of physical device, is
assigned to user space application or a VM. There are cases when multiple
instances of mdev devices of same or different types are required by user
space application or VM. For example in case of vGPU, multiple mdev devices
of type which represents whole GPU can be assigned to one instance of
application or VM.

All types of mdev devices may not support assigning multiple mdev devices
to a user space application. In that case vendor driver can fail open()
call of mdev device. But there is no way to know User space application to
about the configuration supported by vendor driver.

To expose supported configuration, vendor driver should add
'single_usage_restriction' attribute to type-id directory. Returning Y for
this attribute indicates vendor driver has restriction of single mdev
device of particular <type-id> assigned to one user space application.
Returning N indicates that multiple mdev devices of particular <type-id>
can be assigned to one user space application.

User space application should read if 'single_usage_restriction' attibute
is present in <type-id> directory of all mdev devices which are going to be
used. If all read N then user space application can proceed with multiple
mdev devices.

This is optional and readonly attribute.

Signed-off-by: Kirti Wankhede <kwankhede@nvidia.com>
Reviewed-by: Neo Jia <cjia@nvidia.com>
---
 Documentation/ABI/testing/sysfs-bus-vfio-mdev | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/Documentation/ABI/testing/sysfs-bus-vfio-mdev b/Documentation/ABI/testing/sysfs-bus-vfio-mdev
index 452dbe39270e..3aca352a70e5 100644
--- a/Documentation/ABI/testing/sysfs-bus-vfio-mdev
+++ b/Documentation/ABI/testing/sysfs-bus-vfio-mdev
@@ -85,6 +85,22 @@ Users:
 		a particular <type-id> that can help in understanding the
 		features provided by that type of mediated device.
 
+What:		/sys/.../mdev_supported_types/<type-id>/single_usage_restriction
+Date:           October 2018
+Contact:        Kirti Wankhede <kwankhede@nvidia.com>
+Description:
+		Reading this attribute will return Y or N. Returning Y indicates
+		vendor driver has restriction of single mdev device of this
+		particular <type-id> assigned to one user space application.
+		Returning N indicates that multiple mdev devices of particular
+		<type-id> can be assigned to one user space application.
+		This is optional and readonly attribute.
+Users:
+		User space application should read if 'single_usage_restriction'
+		attibute is present in <type-id> directory of all mdev devices
+		which are going to be used. If all read N then user space
+		application can proceed with multiple mdev devices.
+
 What:           /sys/.../<device>/<UUID>/
 Date:           October 2016
 Contact:        Kirti Wankhede <kwankhede@nvidia.com>
